sdk/storage/Azure.Storage.Blobs/src/PartitionedUploader.cs

```diff
@@ -0,0 +1,449 @@
+                    metadata,
+                    conditions,
+                    accessTier,
+                    progressHandler,
+                    cancellationToken)
+                    .ConfigureAwait(false);
+            }
+
+            // If the caller provided an explicit block size, we'll use it.
+            // Otherwise we'll adjust dynamically based on the size of the
+            // content.
+            int blockSize =
+                _blockSize != null ? _blockSize.Value :
+                length < Constants.LargeUploadThreshold ?
+                    Constants.DefaultBufferSize :
+                    Constants.LargeBufferSize;
+
+            // Otherwise stage individual blocks in parallel
+            return await UploadInParallelAsync(
+                content,
+                blockSize,
+                blobHttpHeaders,
+                metadata,
+                conditions,
+                progressHandler,
+                accessTier,
+                cancellationToken)
+                .ConfigureAwait(false);
+        }
+
+        public Response<BlobContentInfo> Upload(
+            Stream content,
+            BlobHttpHeaders blobHttpHeaders,
+            IDictionary<string, string> metadata,
+            BlobRequestConditions conditions,
+            IProgress<long> progressHandler,
+            AccessTier? accessTier = default,
+            CancellationToken cancellationToken = default)
+        {
+            // If we can compute the size and it's small enough
+            if (TryGetLength(content, out long length) && length < _singleBlockThreshold)
+            {
+                // Upload it in a single request
+                return _client.Upload(
+                    content,
+                    blobHttpHeaders,
+                    metadata,
+                    conditions,
+                    accessTier,
+                    progressHandler,
+                    cancellationToken);
+            }
+
+            // If the caller provided an explicit block size, we'll use it.
+            // Otherwise we'll adjust dynamically based on the size of the
+            // content.
+            int blockSize =
+                _blockSize != null ? _blockSize.Value :
+                length < Constants.LargeUploadThreshold ?
+                    Constants.DefaultBufferSize :
+                    Constants.LargeBufferSize;
+
+            // Otherwise stage individual blocks one at a time.  It's not as
+            // fast as a parallel upload, but you get the benefit of the retry
+            // policy working on a single block instead of the entire stream.
+            return UploadInSequence(
+                content,
+                blockSize,
+                blobHttpHeaders,
+                metadata,
+                conditions,
+                progressHandler,
+                accessTier,
+                cancellationToken);
+        }
+
+        private Response<BlobContentInfo> UploadInSequence(
```

Comment:
This one is at least labeled "InSequence" as compared to download, but I still don't understand why we're putting a sequential method in a class labeled "ParallelUploader."

Reply:
The class is called `PartitionedUploader`.  This is both to avoid sync-over-async and needed for arbitrarily large streams.